ha_mroonga.cpp

```diff
@@ -14953,7 +14953,100 @@ mrn_io_and_cpu_cost ha_mroonga::scan_time()
+{
+  MRN_DBUG_ENTER_METHOD();
+  IO_AND_CPU_COST res;
+  MRN_SET_WRAP_SHARE_KEY(share, table->s);
+  MRN_SET_WRAP_TABLE_KEY(this, table);
+  res = wrap_handler->rnd_pos_time(rows);
+  MRN_SET_BASE_SHARE_KEY(share, table->s);
+  MRN_SET_BASE_TABLE_KEY(this, table);
+  DBUG_RETURN(res);
+}
+
+IO_AND_CPU_COST ha_mroonga::storage_rnd_pos_time(ha_rows rows)
+{
+  MRN_DBUG_ENTER_METHOD();
+  IO_AND_CPU_COST res = handler::rnd_pos_time(rows);
+  DBUG_RETURN(res);
+}
+
+IO_AND_CPU_COST ha_mroonga::rnd_pos_time(ha_rows rows)
+{
+  MRN_DBUG_ENTER_METHOD();
+  IO_AND_CPU_COST time;
+  if (share->wrapper_mode)
+  {
+    time = wrapper_rnd_pos_time(rows);
+  } else {
+    time = storage_rnd_pos_time(rows);
+  }
+  DBUG_RETURN(time);
+}
+
+IO_AND_CPU_COST ha_mroonga::wrapper_keyread_time(uint index,
+                                                 uint ranges,
+                                                 ha_rows rows,
+                                                 ulonglong blocks)
+{
+  IO_AND_CPU_COST res;
+  MRN_DBUG_ENTER_METHOD();
+  if (index < MAX_KEY) {
+    KEY *key_info = &(table->key_info[index]);
+    if (mrn_is_geo_key(key_info)) {
+      res = handler::keyread_time(index, ranges, rows, blocks);
+      DBUG_RETURN(res);
+    }
+    MRN_SET_WRAP_SHARE_KEY(share, table->s);
+    MRN_SET_WRAP_TABLE_KEY(this, table);
+    res = wrap_handler->keyread_time(share->wrap_key_nr[index],
+                                     ranges,
+                                     rows,
+                                     blocks);
+    MRN_SET_BASE_SHARE_KEY(share, table->s);
+    MRN_SET_BASE_TABLE_KEY(this, table);
+  } else {
+    MRN_SET_WRAP_SHARE_KEY(share, table->s);
+    MRN_SET_WRAP_TABLE_KEY(this, table);
+    res = wrap_handler->keyread_time(index, ranges, rows, blocks);
+    MRN_SET_BASE_SHARE_KEY(share, table->s);
+    MRN_SET_BASE_TABLE_KEY(this, table);
+  }
+  DBUG_RETURN(res);
+}
+
+IO_AND_CPU_COST ha_mroonga::storage_keyread_time(uint index,
+                                                 uint ranges,
+                                                 ha_rows rows,
+                                                 ulonglong blocks)
+{
+  MRN_DBUG_ENTER_METHOD();
+  IO_AND_CPU_COST time =
+    handler::keyread_time(index, ranges, rows, blocks);
+  DBUG_RETURN(time);
+}
+
+IO_AND_CPU_COST ha_mroonga::keyread_time(uint index,
+                                         uint ranges,
+                                         ha_rows rows,
+                                         ulonglong blocks)
+{
+  MRN_DBUG_ENTER_METHOD();
+  IO_AND_CPU_COST time;
+  if (share->wrapper_mode)
+  {
+    time = wrapper_keyread_time(index, ranges, rows, blocks);
+  } else {
+    time = storage_keyread_time(index, ranges, rows, blocks);
+  }
+  DBUG_RETURN(time);
+}
+#endif
+
+#ifdef MRN_HANDLER_HAVE_READ_TIME
+#  ifdef MRN_ENABLE_WRAPPER_MODE
 double ha_mroonga::wrapper_read_time(uint index, uint ranges, ha_rows rows)
```

Comment:
Can we avoid overriding `read_time` without wrapper mode too?

Reply:
I've fixed it.